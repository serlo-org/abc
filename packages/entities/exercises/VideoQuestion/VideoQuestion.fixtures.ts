import { Fixtures } from '../helpers';
import { IProps, IState } from './VideoQuestion';

const props: IProps = {
  video: require('./../../../../src/assets/videos/placeholder.mp4'),
  question: 'Wer ist eine Ananas?',
  answers: [
    'Ich bin eine Ananas',
    'Du bist eine Ananas',
    'Wir sind eine Ananas'
  ],
  correctIndex: 0
};

const fixtures: Fixtures<IProps, IState> = [
  {
    name: 'correct selected',
    props,
    state: 0,
    isCorrect: true
  },
  {
    name: 'wrong selected',
    props,
    state: 1,
    isCorrect: false
  }
];

export default fixtures;
