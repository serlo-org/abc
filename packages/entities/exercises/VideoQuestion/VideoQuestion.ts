// TODO: don't depend on React Native app
import { IVideoAsset } from '../../../../src/types/assets';
import { Optional } from '../../../../src/types/index';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  video: IVideoAsset;
  question: string;
  answers: string[];
  correctIndex: number;
}

export type IState = Optional<number>;

class VideoQuestion extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return undefined;
  }

  public isCorrect(state: IState) {
    const { correctIndex } = this.getProps();
    return state === correctIndex;
  }
}

export default VideoQuestion;
