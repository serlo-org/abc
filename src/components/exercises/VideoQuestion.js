import React, { Component } from 'react';
import {
  View,
  Text,
  TouchableWithoutFeedback,
  TouchableOpacity,
  Image
} from 'react-native';
import { addIndex, map } from 'ramda';
import { PRIMARY, GREEN } from '../../styles/colors';
import { DEFAULT } from '../../styles/text';
import { RoundTextButton } from '../Components';
import { Video } from 'expo';
import playIcon from '../../assets/images/play.png';

const replayIcon = playIcon;
const mapIndexed = addIndex(map);

const styles = {
  answer: {
    margin: 5
  },
  vidContainer: {
    flex: 3,
    alignItems: 'center',
    justifyContent: 'center'
  },
  video: {
    width: '90%',
    height: '100%'
  },
  button: {
    position: 'absolute',
    padding: 15,
    width: 80,
    height: 80,
    borderRadius: 9999,
    backgroundColor: 'rgba(0, 0, 0, 0.6)',
    zIndex: 9999
  },
  icon: {
    width: 50,
    height: 50
  },
  hidden: {
    width: 0,
    height: 0,
    padding: 0,
    overflow: 'hidden'
  },
  highlighted: {
    backgroundColor: GREEN
  }
};

class VideoQuestion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      highlighted: null,
      isPaused: false,
      isFinished: false
    };
  }

  onEnd = () => {
    this.setState({
      isFinished: true
    });
  };

  playPause = () => {
    this.setState({
      isPaused: !this.state.isPaused
    });
    if (!this.state.isPaused) {
      this.setState({ isFinished: false });
    }
  };

  replay = () => {
    this.player.seek(0);
    this.setState({
      isPaused: false,
      isFinished: false
    });
  };

  selectAnswer = key => () => {
    this.setState(({ highlighted }) => ({
      highlighted: highlighted === key ? null : key
    }));
  };

  render() {
    return (
      <View
        style={{
          alignItems: 'center',
          justifyContent: 'space-around',
          flex: 1,
          backgroundColor: PRIMARY
        }}
      >
        <View style={styles.vidContainer}>
          <TouchableWithoutFeedback onPress={this.playPause}>
            <Video
              ref={ref => {
                this.player = ref;
              }}
              resizeMode={Video.RESIZE_MODE_CONTAIN}
              source={this.props.video}
              style={styles.video}
              paused={this.state.isPaused}
              onEnd={this.onEnd}
            />
          </TouchableWithoutFeedback>
          <TouchableOpacity
            style={[
              styles.button,
              !this.state.isPaused && !this.state.isFinished
                ? styles.hidden
                : null
            ]}
            onPress={this.state.isFinished ? this.replay : this.playPause}
          >
            <Image
              source={this.state.isFinished ? replayIcon : playIcon}
              style={styles.icon}
            />
          </TouchableOpacity>
        </View>
        <View
          style={{ flex: 4, alignItems: 'center', justifyContent: 'center' }}
        >
          <Text style={[DEFAULT, { marginBottom: 25 }]}>
            {this.props.question}
          </Text>
          <View>
            {mapIndexed(
              (item, key) => (
                <RoundTextButton
                  text={item}
                  style={[
                    styles.answer,
                    this.state.highlighted === key ? styles.highlighted : null
                  ]}
                  key={key}
                  onPress={this.selectAnswer(key)}
                />
              ),
              this.props.answers
            )}
          </View>
        </View>
      </View>
    );
  }
}
export default VideoQuestion;
