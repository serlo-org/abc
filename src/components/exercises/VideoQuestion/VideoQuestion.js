import React, { Component } from 'react';
import { View, Text } from 'react-native';
import { addIndex, map } from 'ramda';

import { GREEN } from '../../../styles/colors';
import { DEFAULT } from '../../../styles/text';
import RoundTextButton from '../../common/RoundTextButton';
import Video from '../../common/Video';
import { PortraitScreenOrientation } from '../../helpers/screen-orientation';

const mapIndexed = addIndex(map);

const styles = {
  answer: {
    margin: 5
  },
  vidContainer: {
    flex: 3,
    alignItems: 'center',
    justifyContent: 'center'
  },
  highlighted: {
    backgroundColor: GREEN
  }
};

class VideoQuestion extends Component {
  changeAnswer = key => () => {
    this.props.setState(key);
  };

  render() {
    return (
      <PortraitScreenOrientation>
        <View
          style={{
            alignItems: 'center',
            justifyContent: 'space-around',
            flex: 1
          }}
        >
          <Video video={this.props.video} aspectRatio={3 / 4} />
          <View
            style={{ flex: 4, alignItems: 'center', justifyContent: 'center' }}
          >
            <Text style={[DEFAULT, { marginBottom: 25 }]}>
              {this.props.question}
            </Text>
            <View>
              {mapIndexed(
                (item, key) => (
                  <RoundTextButton
                    text={item}
                    style={[
                      styles.answer,
                      this.props.state === key ? styles.highlighted : null
                    ]}
                    key={key}
                    onPress={this.changeAnswer(key)}
                  />
                ),
                this.props.answers
              )}
            </View>
          </View>
        </View>
      </PortraitScreenOrientation>
    );
  }
}
export default VideoQuestion;
