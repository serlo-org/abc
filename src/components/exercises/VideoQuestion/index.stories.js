import { storiesOf } from '@storybook/react-native';

import * as VideoQuestion from '.';
import { createStoriesFromFixtures } from '../helpers';

const story = storiesOf('exercises/VideoQuestion', module);

createStoriesFromFixtures(story, VideoQuestion);
