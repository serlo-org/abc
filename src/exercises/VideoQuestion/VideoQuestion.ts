import AbstractExercise from '../AbstractExercise';

export interface IProps {
  video: any;
  question: string;
  answers: string[];
  correctIndex: number;
}

export type IState = number | null;

class VideoQuestion extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return null as IState;
  }

  public isCorrect(state: IState) {
    const { correctIndex } = this.getProps();
    return state === correctIndex;
  }
}

export default VideoQuestion;
