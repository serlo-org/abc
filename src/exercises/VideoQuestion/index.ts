import Exercise from './VideoQuestion';
import fixtures from './VideoQuestion.fixtures';

export { Exercise, fixtures };
