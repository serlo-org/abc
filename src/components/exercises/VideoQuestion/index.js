import { VideoQuestion } from '../../../../packages/entities/exercises';
import Component from './VideoQuestion';

export const Exercise = VideoQuestion.Exercise;
export const fixtures = VideoQuestion.fixtures;
export { Component };
