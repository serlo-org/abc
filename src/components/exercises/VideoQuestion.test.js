import React from 'react';
import VideoQuestion from './VideoQuestion';
import renderer from 'react-test-renderer';

jest.mock('expo', () => ({
  Video: 'Video'
}));

it('renders without crashing', () => {
  const tree = renderer.create(
    <VideoQuestion
      video={require('../../assets/videos/placeholder.mp4')}
      question="Wer ist eine Ananas?"
      answers={[
        'Ich bin eine Ananas',
        'Du bist eine Ananas',
        'Wir sind eine Ananas'
      ]}
    />
  );
  expect(tree).toMatchSnapshot();
});
