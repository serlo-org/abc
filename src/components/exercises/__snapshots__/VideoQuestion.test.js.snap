// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00b4d5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 3,
        "justifyContent": "center",
      }
    }
  >
    <Video
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      onEnd={[Function]}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      paused={false}
      resizeMode={undefined}
      source={1}
      style={
        Object {
          "height": "100%",
          "width": "90%",
        }
      }
      testID={undefined}
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.6)",
          "borderRadius": 9999,
          "height": 0,
          "opacity": 1,
          "overflow": "hidden",
          "padding": 0,
          "position": "absolute",
          "width": 0,
          "zIndex": 9999,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Image
        source={1}
        style={
          Object {
            "height": 50,
            "width": 50,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 4,
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          },
          Object {
            "marginBottom": 25,
          },
        ]
      }
    >
      Wer ist eine Ananas?
    </Text>
    <View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#73dbff",
              "borderRadius": 9999,
              "elevation": 5,
              "height": NaN,
              "justifyContent": "center",
              "margin": 5,
              "opacity": 1,
              "padding": 5,
              "shadowColor": "rgba(0, 0, 0, 0.1)",
              "shadowOffset": Object {
                "height": 4,
                "width": 4,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
              "width": NaN,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "norddruck",
                "fontSize": 35,
                "textAlign": "center",
              }
            }
          >
            Ich bin eine Ananas
          </Text>
        </View>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#73dbff",
              "borderRadius": 9999,
              "elevation": 5,
              "height": NaN,
              "justifyContent": "center",
              "margin": 5,
              "opacity": 1,
              "padding": 5,
              "shadowColor": "rgba(0, 0, 0, 0.1)",
              "shadowOffset": Object {
                "height": 4,
                "width": 4,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
              "width": NaN,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "norddruck",
                "fontSize": 35,
                "textAlign": "center",
              }
            }
          >
            Du bist eine Ananas
          </Text>
        </View>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#73dbff",
              "borderRadius": 9999,
              "elevation": 5,
              "height": NaN,
              "justifyContent": "center",
              "margin": 5,
              "opacity": 1,
              "padding": 5,
              "shadowColor": "rgba(0, 0, 0, 0.1)",
              "shadowOffset": Object {
                "height": 4,
                "width": 4,
              },
              "shadowOpacity": 1,
              "shadowRadius": 0,
              "width": NaN,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "norddruck",
                "fontSize": 35,
                "textAlign": "center",
              }
            }
          >
            Wir sind eine Ananas
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
